### PR TITLE
#139 + Upsert refactor

### PR DESCRIPTION
**Pull request by chancancode**
Merged in the latest (4148235) changes in #139 with the upsert refactor: 

Move all the upset code into objects in a new file: each object only need to know how to update within the same type (e.g. html -> html, Node -> Node). If they fail to update because the type has changed, the parent can simply clear the rendered content and make a new upsert object of the appropriate type.

Also fixed CompiledHelpers claiming to correctly return a `PathReference<Opaque>`.

/cc @zackthehuman @krisselden

### CODE REVIEW

File: packages/glimmer-runtime/lib/environment.ts [hunk 1]
```diff
@@ -200,21 +200,31 @@ export abstract class Environment {
...
-export type Insertion = string | SafeString | Node;
+export type Insertion = CautiousInsertion | TrustingInsertion;
+export type CautiousInsertion = string | SafeString | Node;
+export type TrustingInsertion = string | Node;
 
 type PositionalArguments = Opaque[];
 type KeywordArguments = Dict<Opaque>;
 
 export interface Helper {
-  (args: EvaluatedArgs): PathReference<Opaque>;
```
Comment (chancancode) [inline packages/glimmer-runtime/lib/environment.ts]: @zackthehuman is this change intentional? I believe we actually want helpers to return `PathReference`s (see https://github.com/emberjs/ember.js/pull/13173#discussion_r57362608)

File: packages/glimmer-test-helpers/lib/environment.ts [hunk 2]
```diff
@@ -394,7 +394,7 @@ export class SimplePathReference<T> implements PathReference<T> {
 
 type UserHelper = (args: any[], named: Dict<any>) => any;
 
-class HelperReference implements PathReference<Opaque> {
```
Comment (chancancode) [inline packages/glimmer-test-helpers/lib/environment.ts]: See https://github.com/tildeio/glimmer/pull/142/files#r60489824

File: packages/glimmer-test-helpers/lib/environment.ts [hunk 3]
```diff
@@ -409,10 +409,6 @@ class HelperReference implements PathReference<Opaque> {
 
     return helper(positional.value(), named.value());
   }
-
-  get(prop: InternedString): SimplePathReference<Opaque> {
-    return new SimplePathReference(this, prop);
-  }
```
Comment (chancancode) [inline packages/glimmer-test-helpers/lib/environment.ts]: See https://github.com/tildeio/glimmer/pull/142/files#r60489824

File: packages/glimmer-runtime/lib/environment.ts (hunk 1 above)
Comment (zackthehuman) [inline packages/glimmer-runtime/lib/environment.ts]: Yes, otherwise there doesn't seem to be a way to return a ConstReference from a Helper since they aren't PathReferences. Is there a better way to resolve?

Comment (zackthehuman) [inline packages/glimmer-runtime/lib/environment.ts]: We wanted a helper that returned a `ConstReference` so we could write a test for const `SafeStrings` and `Node`s.

Comment (chancancode) [inline packages/glimmer-runtime/lib/environment.ts]: Yep, @krisselden told me about it. There is a few subclasses of `ConstReference` that implements `PathReference`, I'll fix it!

Comment (chancancode) [inline packages/glimmer-runtime/lib/environment.ts]: I was surprised that this change didn't cause any type errors, and I just figured out why. In https://github.com/tildeio/glimmer/blob/master/packages/glimmer-runtime/lib/compiled/expressions/helper.ts#L22, we used `Function.prototype.call` which is typed to be `call(thisArg: any, ...args: any[]): any`, so the return value became `any` and the type system happily let it go through, even though the method is supposed to return a `PathReference<Opaque>` 😞

Fixing that as well

Comment (chancancode) [inline packages/glimmer-runtime/lib/environment.ts]: @zackthehuman @krisselden fixed in 8001311 👍
